src/crimson/common/interruptible_future.h

```diff
@@ -0,0 +1,740 @@
+    : core_type(std::move(fut)) {}
+
+//  [[gnu::always_inline]]
+//  interruptible_future_detail(interruptible_future_detail&& fut) = default;
+
+  ErroratedFuture<::crimson::errorated_future_marker<T...>>
+  to_future() {
+    return static_cast<core_type&&>(std::move(*this));
+  }
+
+  template<typename ValueInterruptCondT, typename ErrorVisitorT>
+  [[gnu::always_inline]]
+  auto safe_then(ValueInterruptCondT&& valfunc, ErrorVisitorT&& errfunc) {
+    auto fut = core_type::safe_then(
+      [func=std::move(valfunc), interrupt_condition=interrupt_cond<InterruptCond>]
+      (T... args) mutable {
+      return call_with_interruption(
+		interrupt_condition,
+		std::move(func),
+		std::forward<T>(args)...);
+      }, [func=std::move(errfunc),
+	  interrupt_condition=interrupt_cond<InterruptCond>]
+	  (auto&& err) mutable {
+	  return call_with_interruption(
+		    interrupt_condition,
+		    std::move(func),
+		    std::move(err));
+      });
+    return (interrupt_futurize_t<decltype(fut)>)(std::move(fut));
+  }
+
+  template <typename ValueInterruptCondT>
+  [[gnu::always_inline]]
+  auto safe_then(ValueInterruptCondT&& valfunc) {
+    auto fut = core_type::safe_then(
+      [func=std::move(valfunc),
+       interrupt_condition=interrupt_cond<InterruptCond>]
+      (T... args) mutable {
+      return call_with_interruption(
+		interrupt_condition,
+		std::move(func),
+		std::forward<T>(args)...);
+    });
+    return (interrupt_futurize_t<decltype(fut)>)(std::move(fut));
+  }
+
+  template <typename ErrorFunc>
+  auto handle_error(ErrorFunc&& errfunc) {
+    auto fut = core_type::handle_error(
+      [errfunc=std::move(errfunc),
+       interrupt_condition=interrupt_cond<InterruptCond>]
+      (auto&& eptr) mutable {
+      return call_with_interruption(
+		interrupt_condition,
+		std::move(errfunc),
+		std::move(eptr));
+    });
+    return (interrupt_futurize_t<decltype(fut)>)(std::move(fut));
+  }
+private:
+  friend class interruptible_future_builder<InterruptCond>;
+  template <typename U>
+  friend struct ::seastar::futurize;
+  template <typename...>
+  friend class ::seastar::future;
+  template<typename TX, typename F>
+  friend inline auto ::seastar::internal::do_with_impl(TX&& rvalue, F&& f);
+  template<typename T1, typename T2, typename T3_or_F, typename... More>
+  friend inline auto ::seastar::internal::do_with_impl(T1&& rv1, T2&& rv2, T3_or_F&& rv3, More&&... more);
+  template <typename T1, typename T2, typename... More>
+  friend auto seastar::internal::do_with_impl(T1&& rv1, T2&& rv2, More&&... more);
+  template <typename HeldState, typename Future>
+  friend class seastar::internal::do_with_state;
+  template <typename, typename>
+  friend class ::crimson::maybe_handle_error_t;
+};
+
+template <typename InterruptCond>
+struct interruptor
+{
+  template <typename... Args>
+  [[gnu::always_inline]]
+  static void enable_interruption(Args&&... args) {
```

Comment:
rename this to "make".

Reply:
@athanatos 
Hi, sam, thanks for your review, I'm adjusting the code as you suggested.

The interrupt condition generally works this way:

1. It is created by "enable_interruption" method, and is recorded in the thread local(I forgot to add "thread_local" to the definition of "interrupt_cond", will add it:-)) global variable "::crimson::interruptible::interrupt_cond";
2. Any continuation that's created within the execution of the continuation that calls the "enable_interruption" method will capture the "interrupt_cond"; and when they starts to run, they will put that capture interruption condition into "::crimson::interruptible::interrupt_cond" so that further continuations created can also capture the interruption condition;
3. At the end of the continuation run, the global "interrupt_cond" will be cleared to prevent other continuations that are not supposed to be interrupted wrongly capture an interruption condition.

I thinks, with this approach, different continuation chains won't be able to interfere with each other as the global "interrupt_cond" only hold their interrupt conditions when their continuations are actually running and no continuations can execute simultaneously.

Each IOInterruptCondition has its own pointer to some specific PG, there can be different interrupt conditions for different PG, I think.

As you can see, the global "interrupt_cond" can work as a signal about whether the continuation is supposed to be interrupted, the reason that I chose to add this global "interrupt_cond" and "enable_interruption" is that there may be this scenario:

- Say there's some method PG::func1(), in which the continuations created may or may not be supposed to be interrupted in different situations, if we don't have a global signal, we have to add an extra parameter to every method like PG::func1() to indicate whether the current run should create to-be-interrupted continuations or not, which is still very error prone.

Are my thoughts right? Thanks:-)

Reply:
I don't see where interrupt_cond is cleared (disable_interruption?).

From what you are saying, I think you mean that

```
PGRef pgref;

using interruptor =
   ::crimson::interruptible::interruptor<
     ::crimson::osd::IOInterruptCondition>;

seastar::future f();
seastar::future g();
seastar::future h();

seastar::future<> run() {
    interruptor::enable_interruption(pgref);
    return f().safe_then([] {
    }).safe_then([] {
        return g();
    }).safe_then([] {
        return h();
    });
}
```

Reply:
@athanatos That's right in general.
But the actual scenario that led to this implementation is like this:
say there's a method  `interruptible_future<> f()`, when called from a client_request continuation chain, the continuations created in it are supposed to be interrupted; however, when invoked in other scenarios they are not supposed to be interrupted. To deal with this scenario, there has to be a signal indicating whether to create an interruption condition for the created continuations, if we don't use a global indicator, we have to add a "bool" parameter to almost every method that may created to-be-interrupted continuations. So I added the global "interrupt_cond". Am I right?:-)

Reply:
How would it work in the following situation?  Wouldn't I need to call enable_interruption in the lambda for the first safe_then?

```
PGRef pgref;

using interruptor =
   ::crimson::interruptible::interruptor<
     ::crimson::osd::IOInterruptCondition>;

seastar::future f();
seastar::future g();
seastar::future h();
seastar::future i();

seastar::future<> run() {
    interruptor::enable_interruption(pgref);
    return f().safe_then([] {
      return g().safe_then([]
        return h();
      }).safe_then([] {
        return i();
      });
  });
}
```

Reply:
Also, do you have an example of the above?  The only ones I can think of would be lower level modules where the continuations shouldn't be interrupted at all.

Reply:
> How would it work in the following situation? Wouldn't I need to call enable_interruption in the lambda for the first safe_then?

Yes, you need to.

> Also, do you have an example of the above? The only ones I can think of would be lower level modules where the continuations shouldn't be interrupted at all.

I think OSDMapGate::wait_for_map can be one, when invoked in from a client_request and referenced from a PG, it might be interrupted, while, in other scenarios, it might not. And further more, I think, there may be possibilities in which different interruption conditions can be created in the same method when it's invoked in different scenarios in the future. So I think adding a global signal might solve this problem once and for all. Is this right?

Reply:
Ok, I think this is a big problem.  While I agree that passing the interrupt condition on every call is a bit error prone, doing it this way makes it incredibly easy for a really small textual refactor that would be fine with normal futures to remove the interrupt condition from a bunch of continuations.   Worse, it's possible for that refactor to have happened in an entirely different file since the relationship is runtime dependent.  The behavior is also dependent on whether operations really block or not during a particular execution, which will make debugging very hard.

Basically, I think we should simply add a safe_then_with_interrupt method to errorator rather than doing all of this.  In most files, we can use a macro to make it somewhat less repetitive.

Reply:
@athanatos "enable_interruption" method has to be called in a continuation that's created by interruptible_future, I guess this is what you are concerned, right? I think maybe I can make it more mandatory:-)

Reply:
Well, you'd have to call enable_interruption for a significant number of the callbacks.  At that point, I'm extremely not convinced that this is better than simply annotating each safe_then call, particularly since with safe_then_with_interrupt it would be easy to tell at a glance whether a particular bit of code is correct during a review.  safe_then_with_interrupt would also be a relatively minimal amount of code which future contributors would have relatively little trouble reading when debugging.  By contrast, this implementation is quite complex and has properties that could hide bugs at run time depending on timing.

Reply:
> Ok, I think this is a big problem. While I agree that passing the interrupt condition on every call is a bit error prone, doing it this way makes it incredibly easy for a really small textual refactor that would be fine with normal futures to remove the interrupt condition from a bunch of continuations. 

I don't quite follow, could you be more specific?

> Worse, it's possible for that refactor to have happened in an entirely different file since the relationship is runtime dependent. The behavior is also dependent on whether operations really block or not during a particular execution, which will make debugging very hard.

I thought seastar tasks won't block, will they?

> 
> Basically, I think we should simply add a safe_then_with_interrupt method to errorator rather than doing all of this. In most files, we can use a macro to make it somewhat less repetitive.

I think the problem is that the interrupt condition in safe_then_with_interrupt has to be provided by callers, which means we have to instantiate an interrupt condition instance for every invocation, which seems no difference from checking the interrupt condition in every continuation which we are trying to avoid.

Reply:
> Well, you'd have to call enable_interruption for a significant number of the callbacks. 

No, enable_interruption needs only be called once, for each continuation chain, at the start of that chain.

Reply:
The case I described above is the example.  Those two pieces of code would behave identically with conventional futures, and transformations like that are common.

seastar tasks don't *block the thread*, but a future may not be ready meaning that the containing task yields control of the cpu back to the reactor so that the reactor can choose another task to run.  I'll generally use the word "block" to mean that a task ceases execution pending some sort of blockage (IO, or a seastar mutex, etc) in this manner.

I don't think the in-memory representation would be any different from what you are doing in your safe_then implementation.  We would capture a reference to a pg-wide object representing the interupt condition in a lambda.

No, enable_interruption must be called *every time a task regains control of the reactor* for the safe_then calls executed prior to the releasing control to know about the interrupt condition.  As I illustrated above in my last example, depending on whether the future returned by f() is ready or not, either the continuations chained to g() will have the interrupt condition or not.  This kind of runtime dependency I claim is a serious debugging problem.

Reply:
> 
> seastar tasks don't _block the thread_, but a future may not be ready meaning that the containing task yields control of the cpu back to the reactor so that the reactor can choose another task to run. I'll generally use the word "block" to mean that a task ceases execution pending some sort of blockage (IO, or a seastar mutex, etc) in this manner.

I think seastar mutex won't block the task that tries to get a lock on it, even if the task can't lock the mutex right away, all the continuations that are supposed to be created, including the ones the creation of which comes after the lock behavior, will be created. So, no matter whether the current task "block", all the continuations that are supposed to be interrupted will capture the interrupt_cond. So I think it won't make any difference about the interrupt behavior. Is this right:-)

> 
> I don't think the in-memory representation would be any different from what you are doing in your safe_then implementation. We would capture a reference to a pg-wide object representing the interupt condition in a lambda.
> 
> No, enable_interruption must be called _every time a task regains control of the reactor_ for the safe_then calls executed prior to the releasing control to know about the interrupt condition. 

I think this job is done in "call_with_interruption" [here](https://github.com/xxhdx1985126/ceph/blob/337e639a2758f57d211120200cea04dc362f2f3f/src/crimson/common/interruptible_future.h#L100), enable_interruption only instantiate the interruption condition, it is the call_with_interruption that ensures each task know about the interrupt condition by setting the global interrupt_cond with the interrupt condition it captured at its creation time.

> The case I described above is the example. Those two pieces of code would behave identically with conventional futures, and transformations like that are common.

I got your point, in your code, if we do safe_then_with_interrupt, we don't need enable_interruption any more. I think the focus of our discussion right now is whether we call safe_then_with_interrupt and instantiate an interrupt condition on every invocation of it or we call enable_interruption once, at the interruptible_future do repetitive work. Is this right?

Reply:
@athanatos Hi, sam, enable_interruption doesn't need to be called every time a task regains control, because every task will invoke call_with_interruption in which it will set the global "interrupt_cond" with its captured "interruption_condition", like I said above.

Reply:
@athanatos The whole procedure is like this:

1. enable_interruption enables interruptions by creating an interrupt condition and set it into "::crimson::interruptible::interrupt_cond"
2. each continuation created afterwards will capture the value of "::crimson::interruptible::interrupt_cond"
3. when those continuations regains control of cpu, it will first set "::crimson::interruptible::interrupt_cond" with its own interrupt condition which is the value of "::crimson::interruptible::interrupt_cond" when the running task is created;
4. the task gets to run
5. the task release the "::crimson::interruptible::interrupt_cond" at the end of its execution

Reply:
Ok, so in my example above, the continuation chained after g() shouldn't actually need to call enable_interruption, right?

Reply:
Yes, that's right:-)